### PR TITLE
[Comunity] Implement Differential Diffusion: Giving Each Pixel Its Strength

### PR DESCRIPTION
**Pull request by huytuong010101**
# Implement Differential Diffusion: Giving Each Pixel Its Strength
- paper: [Differential Diffusion: Giving Each Pixel Its Strength](https://arxiv.org/abs/2306.00950)
- readme: update after PR is accept

### CODE REVIEW

**Comment by noskill** [comment]
the code looks very similar to masked im2im pipeline https://github.com/huggingface/diffusers/blob/main/examples/community/masked_stable_diffusion_img2img.py

i didn't add noise to decoded latents in this equation:

image

because results where the same as without noise, but it seems the only difference. @huytuong010101 what do you think about adding this functionality into masked im2im pipeline?

**Comment by noskill** [comment]
@huytuong010101 in MaskedStableDiffusionImg2ImgPipeline mask is also non-binary, that's why i suggest to merge your changes into MaskedStableDiffusionImg2ImgPipeline

The only difference i see is here:

latents = latents * (1 - step_mask) + latents_each_step[i] * (step_mask)

while in my implementation it's always latents_each_step[0]

so my suggestion is to add this as an option into MaskedStableDiffusionImg2ImgPipeline, maybe with additional argument in *\_\_call\_\_*, something like that: **use_denoise_steps=True**

**Comment by recoilme** [comment]
@noskill @huytuong010101 
May pls provide some example in both implementations (mask and image)?

**Comment by noskill** [comment]
@recoilme sure:



```
import PIL
import torch
import numpy
import diffusers
from diffusers import *

img = PIL.Image.open("./mech.png")
mask = PIL.Image.open("./mask_gradient.png")
mask = numpy.array(mask)[:, :, 0]
mask = mask / mask.max()

pipeline = StableDiffusionImg2ImgPipeline.from_pretrained("frankjoshua/icbinpICantBelieveIts_v8", custom_pipeline="masked_stable_diffusion_img2img")
pipeline.to('cuda')

# works best with EulerAncestralDiscreteScheduler
pipeline.scheduler = EulerAncestralDiscreteScheduler.from_config(pipeline.scheduler.config)
generator = torch.Generator(device="cpu").manual_seed(4)

prompt = "a man wearing a mask"
result = pipeline(prompt=prompt, image=img, mask=mask, strength=1,
                          generator=generator)
result.images[0].save("result.png")
```
mech.png

mask_gradient.png


result.png

**Comment by recoilme** [comment]
@noskill thank you for example!
This may be the wrong topic, but I'm getting bad results compared to the auto1111 mask implementation.

In auto1111 mask implemented as transparency percent:

original image
image2

mask with transparency (gradient from face)
result_mask

result in auto1111 - (photo of a man, balensiaga style)
masked

You use rgb image with inverted mask where white color - max difference, so i invert/convert to jpg
result_mask2

And get this with this version
result

Looks like auto1111 has very different (and more correct by my opinion) implementation

Or may be i do something wrong?
My code very simple, only different part is 
```
prompt = "photo of a man, balensiaga style"
negative_prompt = "deformed, disfigured, poorly drawn, bad anatomy, wrong anatomy, limb, mutated, blurry"
result = pipeline(num_inference_steps=10,negative_prompt=negative_prompt, prompt=prompt, image=img, mask=mask, strength=1,generator=generator)
```
I play with strength / num_inference_steps 0.3-0.7, 10/15/20 -  result less terrible (just blurred), but not comparable to auto1111 implementation quality(

(i am on mps backend with PYTORCH_ENABLE_MPS_FALLBACK=1)

other points that seemed unobvious:  i think transparency percent more understandable, it will be prefect to support rgba images to (bgr and so on), add example in method, and use plain array as percent of strength

**Comment by noskill** [comment]
@recoilme I got better results with increased brightness of the mask:
mask_bright
result

Also i converted  mask to greyscale 

```
import PIL
import torch
import numpy
import diffusers
from diffusers import *
from community.masked_stable_diffusion_img2img import MaskedStableDiffusionImg2ImgPipeline

img = PIL.Image.open("./input.jpeg")
mask = PIL.Image.open("./mask_bright.jpg")
mask = numpy.array(mask)
mask = mask / mask.max()

models_dir = "/home/noskill/models/"
model_id = "icbinpICantBelieveIts_v8.safetensors"
pipeline = MaskedStableDiffusionImg2ImgPipeline.from_single_file(models_dir + model_id)
pipeline.to('cuda')

# works best with EulerAncestralDiscreteScheduler
pipeline.scheduler = EulerAncestralDiscreteScheduler.from_config(pipeline.scheduler.config)
generator = torch.Generator(device="cpu").manual_seed(4)

prompt = "photo of a man, balensiaga style"
negative_prompt = "deformed, disfigured, poorly drawn, bad anatomy, wrong anatomy, limb, mutated, blurry"

result = pipeline(prompt=prompt, image=img, mask=mask, strength=0.8,
                          generator=generator)
result.images[0].save("result.png")
```


There are artifacts in the area where the headphones are, there you need to add stronger mask perhaps?
Also automatic1111 has a few different options how to initialise latents in the masked area: random, zeroes, from image/from mask. Also automatic blends original image with inpainted one using blurred mask and also there are some resizes.. I didn't port everything yet.

with 10 inference steps the results indeed very blurry

I'll check automatic111 implementation  with your images though, i maybe i missed something important

btw did you use sketch inpaint or just image to image?

**Comment by recoilme** [comment]
> btw did you use sketch inpaint or just image to image?
i use image 2 image / inpaint upload / with 0.5 denoising strength / (euler a or DPM++ 2m Karras)

I use this script for mask generation: https://gist.github.com/recoilme/a6f24ac83e89a0cecea9ba662d7f5e1a
(i'm not python dev, script writen with chatgpt support, sorry for that)

Problem not in this photo, i done many photo convertations with mask in auto1111 and result not looking blurry
Снимок экрана 2023-10-02 в 12 05 45
Снимок экрана 2023-10-02 в 12 06 07

Probably (i'm not in python, sorry), auto1111 has some rounding inside, and 0.8 weight coverts to 1
But i dont may create enough tests on my mac m1 (slow), or understand difference by looking at code(

But then i use mask 0 .. 1 i get contour in auto1111, that's why I think rounding is used

**Comment by noskill** [comment]
@recoilme ok, i checked automatic111 implementation. It does a few things that MaskedStableDiffusionImg2ImgPipeline doesn't.

1. Fist mask gets blurred

2. Then, and this is the most important is this peace of code:

np_mask = np.clip((np_mask.astype(np.float32)) * 2, 0, 255).astype(np.uint8)

So in fact  the mask gets much more bright. 

3. Then they store overlay image, they merge it later with the generated one:

It was in rgba format, here is what i got when converted it to rgb

image_masked

I have code for blurring the mask and merging generated image with original image here 

https://github.com/singnet/metafusion/blob/main/multigen/pipes.py#L212 

I am not sure this mask multiplication and clipping is that important, the next thing i wanted to do is to make the same pipeline work with inpainting models. But i guess we could add this as an option.

**Comment by recoilme** [comment]
@noskill omg, so tricky guys in automatic)
Ty so much for your investigation and time!
Agree, it's must have feature for smooth inpainting by mask (but it more about tricks then each pixel with different strength)

**Comment by adhikjoshi** [comment]
> @recoilme ok, i checked automatic111 implementation. It does a few things that MaskedStableDiffusionImg2ImgPipeline doesn't.
> 
> 
> 
> 1. Fist mask gets blurred
> 
> 
> 
> 2. Then, and this is the most important is this peace of code:
> 
> 
> 
> np_mask = np.clip((np_mask.astype(np.float32)) * 2, 0, 255).astype(np.uint8)
> 
> 
> 
> So in fact  the mask gets much more bright. 
> 
> 
> 
> 3. Then they store overlay image, they merge it later with the generated one:
> 
> 
> 
> It was in rgba format, here is what i got when converted it to rgb
> 
> 
> 
> image_masked
> 
> 
> 
> I have code for blurring the mask and merging generated image with original image here 
> 
> 
> 
> https://github.com/singnet/metafusion/blob/main/multigen/pipes.py#L212 
> 
> 
> 
> I am not sure this mask multiplication and clipping is that important, the next thing i wanted to do is to make the same pipeline work with inpainting models. But i guess we could add this as an option.

This looks much better, imo it needs to be added also in other inpaiting pipelines.

**Comment by noskill** [comment]
@adhikjoshi do you mean this 2x multiplication + clipping or overlaying with original image?

**Comment by adhikjoshi** [comment]
> @adhikjoshi do you mean this 2x multiplication + clipping or overlaying with original image?

Yes, this gives better quality.
